### PR TITLE
SKoutpost directive and overload: improvements and fixes

### PR DESCRIPTION
**Pull request by zGeneral**
## Pull request summary

Edit: verified and checked again on 14/June, waiting for PR review.

(all tested and working properly now)!

SKOuptpost container construction sites for minerals are now pushed to colony worker build list.
similarly,
SKOuptpost containers with hits < 50% are now maintained by workers (previously, harvesters maintain their own containers, but mineral drones have their container unrepaired and decay to destruction)
 
When there is no container:
drones extracting minerals assume containers already exist to transfer mineral in.
when there is no container (mainly in SK rooms due to container build delay) the drones stay idle and die of age.
this PR makes them travel back to colony for to drop it off as a temp alternative option until the container is built.

this should also fix issue #120, miners spawn in own room without having terminal/storage

sourceReapers now do not spawn in center SK room
### Added:

- None

### Changed:

- None

### Removed:

- None

### Fixed:

issue #120 

- None

## Testing checklist:

- [x] Changes are backward-compatible OR version migration code is included
- [x] Codebase compiles with current `tsconfig` configuration
- [x] Tested changes on *{choose PUBLIC/PRIVATE}* server OR changes are trivial (e.g. typos)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,14 +31,15 @@
     "rollup-plugin-node-resolve": "3.3.0",
     "rollup-plugin-progress": "^0.4.0",
     "rollup-plugin-screeps": "0.1.2",
-    "rollup-plugin-typescript2": "0.16.1",
+    "rollup-plugin-typescript2": "^0.21.1",
```
Comment (bencbartlett) [inline package.json]: I'm hesitant to make changes to dependencies, as it can break the checksum generation process. Is there a reason you are updating the rollup plugin?

File: src/directives/colony/outpostSK.ts [hunk 2]
```diff
@@ -12,21 +15,74 @@ export class DirectiveSKOutpost extends Directive {
...
 	static secondaryColor = COLOR_YELLOW;
+	static containerRepairTheshold = 0.5;
+
+	memory: DirectiveSKOutpostMemory;
 
 	constructor(flag: Flag) {
 		super(flag, colony => colony.level >= 7);
 	}
+	
+	isCenterRoom(): boolean {
```
Comment (bencbartlett) [inline src/directives/colony/outpostSK.ts]: this is unnecessary; use something like `const isCenterRoom = Cartographer.roomType(this.pos.roomName) == ROOMTYPE_CORE`

File: src/directives/colony/outpostSK.ts [hunk 3]
```diff
@@ -12,21 +15,74 @@ export class DirectiveSKOutpost extends Directive {
...
+		if(coords.x % 10 != 0 && coords.x % 5 == 0 && coords.y % 10 != 0 && coords.y % 5== 0){
+			return true;
+		}
+		return false;
+	}
 
 	spawnMoarOverlords() {
-		this.overlords.sourceReaper = new SourceReaperOverlord(this);
+		if(this.memory.isCenterRoom == undefined){
+			this.memory.isCenterRoom = this.isCenterRoom();
```
Comment (bencbartlett) [inline src/directives/colony/outpostSK.ts]: don't store `isCenterRoom` in memory. it's inexpensive to calculate and can be stuck directly in the DirectiveSKOutpost constructor

File: src/directives/colony/outpostSK.ts [hunk 4]
```diff
@@ -12,21 +15,74 @@ export class DirectiveSKOutpost extends Directive {
...
+		}			
+		return;
 	}
 
 	init(): void {
+		// Add this structure/CS to worker overlord's build/repair list
+		if(Game.time % 150 == 0){
+			const containerNeedRepair = this.getContainersToRepair();
+			if (containerNeedRepair && !this.colony.overlords.work.repairStructures.includes(containerNeedRepair)) {
+				this.colony.overlords.work.repairStructures.push(containerNeedRepair);
```
Comment (bencbartlett) [inline src/directives/colony/outpostSK.ts:59]: sk miners already build and repair their own containers -- why do you want to offload this to the colony workers?

File: src/directives/resource/extract.ts [hunk 5]
```diff
@@ -36,7 +36,10 @@ export class DirectiveExtract extends Directive {
 		} else {
 			priority = OverlordPriority.remoteSKRoom.mineral;
 		}
-		this.overlords.extract = new ExtractorOverlord(this, priority);
+		//Fix: Only spawn drones if there is a terminal or storage
+		if (!!this.colony.terminal || !!this.colony.storage){
```
Comment (bencbartlett) [inline src/directives/resource/extract.ts]: good catch

File: src/overlords/mining/extractor.ts [hunk 6]
```diff
@@ -105,6 +105,16 @@ export class ExtractorOverlord extends Overlord {
 	}
 
 	private handleDrone(drone: Zerg): void {
+		//fix: if there is no container, then transfer the minerals yourself!
+		if(!this.container && _.sum(drone.carry) == drone.carryCapacity ){
```
Comment (bencbartlett) [inline src/overlords/mining/extractor.ts]: we probably just shouldn't spawn any extractors unless there is a container

File: src/overlords/mining/sourceKeeperReeper.ts [hunk 7]
```diff
@@ -95,7 +95,7 @@ export class SourceReaperOverlord extends CombatOverlord {
 				reaper.healSelfIfPossible();
 			}
 			// Kite around ranged invaders until a defender arrives
-			if (this.room.invaders.length > 2 && _.filter(this.defenders, def => def.room == this.room).length == 0) {
+			if (this.pos.findInRange(this.room.invaders,10).length > 2 && _.filter(this.defenders, def => def.room == this.room).length == 0) {
```
Comment (bencbartlett) [inline src/overlords/mining/sourceKeeperReeper.ts]: why this change?

**Review by bencbartlett — CHANGES_REQUESTED** [review]
see comments -- there are some good changes in here but also some things I'm confused about

File: src/directives/colony/outpostSK.ts (hunk 4 above)
Comment (zGeneral) [reply src/directives/colony/outpostSK.ts:59]: drones that harvest minerals do not repair their containers like the drones harvesting energy.
I have noticed in replays, that mineral containers always decays out without repair.

File: src/overlords/mining/sourceKeeperReeper.ts (hunk 7 above)
Comment (zGeneral) [reply src/overlords/mining/sourceKeeperReeper.ts]: will revert this one back.

File: src/overlords/mining/extractor.ts (hunk 6 above)
Comment (zGeneral) [reply src/overlords/mining/extractor.ts]: good idea, 
what about changing the amount to zero in wishlist if there is no container, something like this:
const amount = this.mineral && this.mineral.mineralAmount > 0 **&& this.container** ? this.mineral.pos.availableNeighbors().length : 0;

File: src/directives/colony/outpostSK.ts (hunk 4 above)
Comment (zGeneral) [reply src/directives/colony/outpostSK.ts:59]: this changes does not have a filter for mineral container only, the worker will repair all containers in SKroom below 50%, which will most probably apply to mineral container being below the threshold

File: package.json (hunk 1 above)
Comment (zGeneral) [reply package.json]: fixed, check latest PR

File: src/directives/colony/outpostSK.ts (hunk 2 above)
Comment (zGeneral) [reply src/directives/colony/outpostSK.ts]: done, check latest PR

File: src/directives/colony/outpostSK.ts (hunk 3 above)
Comment (zGeneral) [reply src/directives/colony/outpostSK.ts]: done, it is now checked in constructor.
existing Overmind users that already have the directive flag in place will not benefit from this, would you suggest to add it under init() too?
